File: sfes.py
import os
import sys
from getopt import *

class bcolors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKCYAN = '\033[96m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'

def colored(text, color):
	match color:
		case "header":
			return bcolors.HEADER+text+bcolors.ENDC
		case "okblue":
			return bcolors.OKBLUE+text+bcolors.ENDC
		case "okcyan":
			return bcolors.OKCYAN+text+bcolors.ENDC
		case "okgreen":
			return bcolors.OKGREEN+text+bcolors.ENDC
		case "warning":
			return bcolors.WARNING+text+bcolors.ENDC
		case "fail":
			return bcolors.FAIL+text+bcolors.ENDC
		case "bold":
			return bcolors.BOLD+text+bcolors.ENDC
		case "underline":
			return bcolors.UNDERLINE+text+bcolors.ENDC
		case _:
			return text

def sort(prefix):
	files = os.listdir()
	extensions = list()
	for file in files:
		if "." in file and os.path.basename(__file__) != file and os.path.isfile(file) and file != "desktop.ini" and file != "cfg.sfes":
			extensions.append(file.split(".")[-1])

	extensions = list(dict.fromkeys(extensions))
	for ext in extensions:
		try:
			os.mkdir(f"{prefix}{ext}")
		except:
			pass
	for file in files:
		for ext in extensions:
			if file.split(".")[-1] == ext:
				try:
					os.rename(file, f"{prefix}{ext}/{file}")
				except Exception as e:
					print(colored(f"Failed {file} to move: {e}", "fail"))
	print(colored("Done!", "okgreen"))
	with open("cfg.sfes", "w") as file:
		file.write(prefix)
	print(colored("Do not delete or edit cfg.sfes file if you will need to unsort your files!", "warning"))
	sys.exit()


def unsort(prefix):
	try:
		with open("cfg.sfes") as file:
			prefix = file.readlines()[0]
	except FileNotFoundError:
		print(colored("Error: cfg.sfes not found, cannot resort", "fail"))
		sys.exit()
	files = os.listdir()
	folders = list()
	extensions = list()
	for file in files:
		if prefix in file and os.path.isdir(file):
			folders.append(file)
			extensions.append(file.split(".")[-1])
	for folder in folders:
		files = os.listdir(folder)
		for file in files:
			try:
				os.rename(f"{folder}/{file}", f"{file}")
			except Exception as e:
				print(colored(f"Failed {file} to move: {e}", "fail"))
		try:
			os.rmdir(folder)
		except Exception as e:
			print(colored(f"Failed {file} to delete: {e}", "fail"))
	print(colored("Done!", "okgreen"))
	sys.exit()


def help():
	print(colored("Usage:\n-p (prefix) - sets prefixes for folders like '(prefix)txt' or '(prefix)doc'\n-n - create folders named by just extensions\n-r - sorts unsorted files in existant folders using previous prefix\n-h - shows this message\n-c - unsort, bring files back from folders (also deleting folders)", "warning"))
	sys.exit()


def main():
	print(colored("Simple File Extension Sorter v1.12\n", "header"))
	if len(sys.argv) == 1:
		help()
	try:
		optlist, args = getopt(sys.argv[1:], 'np:hcr')
	except GetoptError as e:
		print(colored("Error: " + str(e) + ", use 'sfes -h' for help", "fail"))
		sys.exit()

	for opt, a in optlist:
		if opt == "-h":
			help()
		elif opt == "-n":
			yn = input(colored("Are you sure that you want to have no prefix? You wont be able to unsort your files using sfes in the future!\n(yes or no? y/n): ", "warning"))
			if yn.lower() == "y":
				sort("")
			elif yn.lower() == "n":
				sys.exit()
			else:
				print("I`ll take that as no")
				sys.exit()
		elif opt == "-p":
			if a not in "+=[]:;«,./\\:*? «<>|":
				sort(a)
			else:
				print(colored("Error: Can't create folders with this prefix!", "fail"))
				sys.exit()
		elif opt == "-c":
			unsort(a)
		elif opt == "-r":
			try:
				with open("cfg.sfes") as file:
					sort(file.readlines()[0])
			except FileNotFoundError:
				print(colored("Error: cfg.sfes not found, cannot resort", "fail"))
				sys.exit()

	


if __name__ == "__main__":
	main()

File: sfes/sfes.cpp
#include <filesystem>
#include <iostream>
#include <string>
#include <fstream>

void sort(char* prefix)
{
    //  TODO: бл€€€€ть сука ебучие массивы в си++ короче сука
    std::vector<std::string> entities;
    std::ofstream ConfigFile("cfg.sfes");
    ConfigFile << prefix;
    ConfigFile.close();
    std::cout << "This doesn't work yet, prefix is '" << prefix << "'" << std::endl;
    std::string path = "./";
    for (const auto& entry : std::filesystem::directory_iterator(path))
    {
        entities.push_back(entry.path().u8string());
        //std::cout << entry.path().u8string() << std::endl;
    }
    for (int i = 0; i < entities.size(); i++)
    {
        std::cout << entities[i] << std::endl;
    }

}

void help(void)
{
    /*
    std::cout << "Usage:" << std::endl
        << " -p (prefix) - sets prefixes for folders like '(prefix)txt' or '(prefix)doc'" << std::endl
        << " -n - create folders named by just extensions" << std::endl
        << " -r - sorts unsorted files in existant folders using previous prefix" << std::endl
        << " -h - shows this message" << std::endl
        << " -c - unsort, bring files back from folders(also deleting folders)" << std::endl; */

    std::cout << "help message duh" << std::endl;
    return;
}

int main(int argc, char* argv[])
{
    std::cout << "Simple File Extension Sorter (rewritten)" << std::endl;
    if (argc > 1)
    {
        if (argv[1][1] == 'p' && argc == 3)
        {
            std::cout << "PLEAEEASE:3WORK ALREADY:3" << std::endl;
            sort(argv[2]);
        }
        else if (argv[1][1] == 'n')
        {
            sort(0);
        }
        else if (argv[1][1] == 'r')
        {
            std::ifstream cfg("cfg.sfes");
            if (cfg.good())
            {
                std::string prefix;
                std::getline(cfg, prefix);
                cfg.close();
                sort(prefix.data());
            }
            else
            {
                std::cout << "No previous sorts found! (No cfg.sfes file in the directory)";
            }

        }
        else if (argv[1][1] == 'c')
        {
            //not yet
            help();
        }
        else {
            help();
        }
    }    
    else {
        help();
    }
    return 1;
}
